packages/gatsby/src/internal-plugins/internal-data-bridge/gatsby-node.js

```diff
@@ -41,15 +40,17 @@ function transformPackageJson(json) {
 
 const createPageId = path => `SitePage ${path}`
 
-exports.sourceNodes = ({ createContentDigest, actions, store }) => {
-  const { createNode } = actions
+exports.sourceNodesStatefully = ({ createContentDigest, actions, store }) => {
+  const { createNode, deleteNode } = actions
   const state = store.getState()
   const { program } = state
   const { flattenedPlugins } = state
```

Comment:
One issue is that we have listeners below that will capture this `state` and since we never "unlisten" them they will prevent garbage collection of everything in this state. We should probably not reference state at all and just destructure `program` and `flattenedPlugins` (and maybe reset every variable that is not used by listeners when it is not needed anymore).

P.S. Initially I thought that JS engines are smart enough to detect unused variables and exclude them from captured closure context but it doesn't seem to be the case (at least not always)

Reply:
Alternatively I can move this stuff around so `state` is not available in long-term listener closures so it can be safely garbage collected. 

Will see what I can come up with. Thanks for tracking leak problem here!